File: leetcode090.py
class Solution:
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = [[]]
        begin = 0
        for i in range(len(nums)):
            if i == 0 or nums[i] != nums[i-1]:
                begin = 0
            size = len(res)
            for j in range(begin,size):
                curr = list(res[j])
                curr.append(nums[i])
                res.append(curr)
            begin = size

        return res


if __name__ == "__main__":
    s = Solution()
    print(s.subsetsWithDup([1,2,2]))

File: leetcodec++/leetcode090.cpp
//
// Created by 黄敏辉 on 2019-07-30.
//

class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> comb;
        helper(res, nums, comb, 0);
        return res;
    }

    void helper(vector<vector<int>>& res, vector<int>& nums, vector<int>& comb, int begin){
        res.push_back(comb);
        for (int i = begin; i<nums.size();i++){
            if (i ==begin || nums[i] != nums[i-1]){
                comb.push_back(nums[i]);
                helper(res, nums, comb, i+1);
                comb.pop_back();
            }
        }
    }
};